spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaListenerAnnotationBeanPostProcessor.java

```diff
@@ -411,8 +412,23 @@ public Object postProcessAfterInitialization(final Object bean, final String bea
 					Set<Method> methodsWithHandler = MethodIntrospector.selectMethods(targetClass,
 							(ReflectionUtils.MethodFilter) method ->
 									AnnotationUtils.findAnnotation(method, KafkaHandler.class) != null);
-					List<Method> multiMethods = new ArrayList<>(methodsWithHandler);
-					processMultiMethodListeners(classLevelListeners, multiMethods, targetClass, bean, beanName);
+					Set<Method> publicMethods = MethodIntrospector.selectMethods(targetClass,
+							(ReflectionUtils.MethodFilter) method ->
+									Modifier.isPublic(method.getModifiers()));
+
+					if (methodsWithHandler.isEmpty() && publicMethods.size() == 1 && !hasMethodLevelListeners) {
+						// Case when target class has class-level @KafkaListener annotation and
+						// has only single public method without @KafkaHandler.
+						// See, GH-3807 for more details.
```

Comment:
I don't think we need this comment here. The behavior is clear from the logic of these `if..else`